proposals/0000-grand-class-unification.rst

```diff
@@ -0,0 +1,167 @@
+The Grand Plan for Unification of Class Methods and Associated Types
+====================================================================
+
+.. proposal-number::
+.. ticket-url::
+.. implemented::
+.. highlight::
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/236>`_.
+.. sectnum::
+.. contents::
+
+NB. This is a "big picture" proposal, prompted by several other threads on the
+GHC Proposals platform. It is not intended to be submitted to the committee in
+its current form; rather, it outlines a general direction of travel for other,
+less ambitious proposals, to take into account.
+
+The proposed features are:
+
+* top-level signatures for class methods
+* remove the ``type`` prefix of top-level kind signatures
+* visible dependent quantification in terms
+* visible quantification of ambiguous class variables
+* constrained type families
+
+In more detail:
+
+* Add top-level signatures for class methods that serve as machine-checked
+  documentation::
+
+    class Eq a where
+      eq :: a -> a -> a
+
+    eq :: forall t. Eq t => t -> t -> t
+
+  The rule for checking the top-level signature is as follows. Let us refer to
+  the in-class method as ``$eq``, then we assume the following definition::
+
+    eq = $eq
+
+  This allows us to reorder and rename invisible type variables, and to
+  use a subclass instead of the actual parent class.
+
+* Remove the ``type`` prefix of top-level kind signatures::
+
+    Functor :: (Type -> Type) -> Constraint
+    class Functor f where
+      fmap :: (a -> b) -> f a -> f b
+
+  This unifies TLKS and regular, term-level signatures. In case of namespace
+  ambiguity, reject::
+
+    data T = T
+    T :: ...   -- not allowed.
+
+* Allow visible dependent quantification in terms::
+
+    idv :: forall a -> a -> a
+    idv t x = x :: t
+
+    x = idv Double  42     -- x :: Double
+    n = idv Integer 42     -- n :: Integer
+
+  The difference with ``-XTypeApplications`` is in the presence of a visibility
+  override ``@`` at use sites.  This implies a unification of the type and
+  expression parsers, and a change to name lookup rules.
+
+* Quantify class variables visibly when ambiguous::
```

Comment:
As far as I can tell, this is a completely new idea. As a result, I think this deserves more justification for why it is the right thing to do. Personally, I'm somewhat suspicious of this idea, since it's very easy to write Haskell98 class declarations that are ambiguous, and having those declarations imply visible dependent quantification feels like a step too far (at least, without some sort of language extension to guard it). But perhaps I'm simply not seeing all of the details of the design you imagine for this feature.

Reply:
> it's very easy to write Haskell98 class declarations that are ambiguous, and having those declarations imply visible dependent quantification feels like a step too far 

Haskell 98 and Haskell 2010 reject these ambiguous declarations. We would fail with an error about ambiguous type variables unless either `-XVDQ` is enabled.

Reply:
> unless either `-XVDQ` is enabled.

Ah, I had no idea you intended to have a dedicated language extension for this! That's a very important detail to mention.